### PR TITLE
feat(linux): make deep link auth work

### PR DESCRIPTION
**Pull request by ReactorScram**
Right now it only works on my dev VM, not on my test VMs, due to #4053 and #4103, but it passes tests and should be safe to merge.

There's one doc fix and one script fix which are unrelated and could be their own PRs, but they'd be tiny, so I left them in here.

Ref #4106 and #3713 for the plan to fix all this by splitting the tunnel process off so that the GUI runs as a normal user.

### CODE REVIEW

File: rust/gui-client/build.sh [hunk 1]
```diff
@@ -1,18 +1,15 @@
-#!/bin/sh
+#!/usr/bin/env bash
```
Comment (ReactorScram) [inline rust/gui-client/build.sh:1]: all this stuff is unrelated to deep linking but happened to be changed on this branch.

File: rust/gui-client/docs/manual_testing.md [hunk 2]
```diff
@@ -89,6 +91,10 @@ This is a list of all the on-disk state that you need to delete / reset to test
 - Registry key `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Tcpip\Parameters\Interfaces\{e9245bc1-b8c1-44ca-ab1d-c6aad4f13b9c}` (IP address and DNS server for our tunnel interface)
 - Windows Credential Manager, "Windows Credentials", "Generic Credentials", `dev.firezone.client/token`
 
+## Linux
+
+- Dir `$HOME/.local/share/applications` (.desktop file for deep links. This dir may not even exist by default on distros like Debian)
+
```
Comment (ReactorScram) [inline rust/gui-client/docs/manual_testing.md:97]: Not complete yet but it's more than nothing.

File: rust/gui-client/src-tauri/src/client/deep_link.rs [hunk 3]
```diff
@@ -26,42 +26,24 @@ mod imp;
...
 pub(crate) use imp::{open, register, Server};
 
-pub(crate) fn parse_auth_callback(url: &SecretString) -> Option<AuthResponse> {
-    let url = Url::parse(url.expose_secret()).ok()?;
-    match url.host() {
-        Some(url::Host::Domain("handle_client_sign_in_callback")) => {}
-        _ => return None,
+pub(crate) fn parse_auth_callback(url: &SecretString) -> Result<AuthResponse> {
+    // TODO: Delete this before opening PR
+    tracing::debug!(secret = url.expose_secret(), "Parsing URL");
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/deep_link.rs]: oops

File: rust/gui-client/src-tauri/src/client/deep_link/linux.rs [hunk 4]
```diff
@@ -1,29 +1,121 @@
...
+pub(crate) struct Server {
+    listener: UnixListener,
+}
 
 impl Server {
-    pub(crate) fn new() -> Result<Self, Error> {
-        tracing::warn!("Not implemented yet");
-        tracing::trace!(scheme = super::FZ_SCHEME, "prevents dead code warning");
-        Ok(Self {})
+    /// Create a new deep link server to make sure we're the only instance
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/deep_link/linux.rs:16]: Single-instance doesn't work on Linux yet. It's in #3884

File: rust/gui-client/src-tauri/src/client.rs [hunk 5]
```diff
@@ -212,6 +217,7 @@ pub enum Cmd {
 
 #[derive(Args)]
 pub struct DeepLink {
+    // TODO: Should be `Secret`?
```
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client.rs:220]: Yeah, but we can wait until the two-process arch is done before this

File: rust/gui-client/src-tauri/src/client/deep_link.rs [hunk 6]
```diff
@@ -26,42 +26,22 @@ mod imp;
...
 
 pub(crate) use imp::{open, register, Server};
 
-pub(crate) fn parse_auth_callback(url: &SecretString) -> Option<AuthResponse> {
-    let url = Url::parse(url.expose_secret()).ok()?;
-    match url.host() {
-        Some(url::Host::Domain("handle_client_sign_in_callback")) => {}
-        _ => return None,
+pub(crate) fn parse_auth_callback(url: &SecretString) -> Result<AuthResponse> {
+    let url = Url::parse(url.expose_secret())?;
```
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/deep_link.rs]: Is url a Secret or Url?

**Review by jamilbk — APPROVED** [review]

File: rust/gui-client/src-tauri/src/client/deep_link.rs (hunk 6 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/deep_link.rs]: hm looks like I shadowed it. I'll rename it real quick for clarity before merging

**Pull request merged**